add Jetson Orin support

thomas-hiddenpeak opened the pull request:
# Motivation and Context
NVIDIA Jetson Orin devices have a compute capability of 8.7, which is not currently supported in the compute_cap_matching function. This PR ensures that these devices can be used with the library by adding the necessary support.

# What does this PR do?

This PR adds support for NVIDIA Jetson Orin devices by including the compute capability 8.7 in the `compute_cap_matching` function and updating the tests to ensure the new capability is correctly supported.

Fixes #466

Checklist
 - [x] I have read the [contributor guidelines](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests).
 - [x] I have added tests to verify my changes.
 - [x] I have tagged the appropriate reviewers.

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag members/contributors who may be interested in your PR.

@OlivierDehaene OR @Narsil

r0kk commented:
@HiddenPeak I am wondering if you could share reproducible steps how were you able to run text-embeddings-inference on Jetson AGX Orin. It would be greatly appreciated🙏.

Unfortunately I don't have deep enough knowledge to review you PR.

HiddenPeak commented:
The Jetson Orin series uses the CUDA architecture SM8.7, which is part of the Ampere architecture. Theoretically, it should be compatible with TEI. However, in practical applications, there are many incompatibilities, making direct support generally impossible. During the process of attempting to use it, I encountered the following issues:

1. The compute_cap_matching() function does not support the SM87 architecture, so I modified the source code and recompiled it.
2. It is necessary to ensure that the GPU driver, CUDA runtime, and CUDA compiler are correctly installed and available in the environment variable paths.（on Jetpack 6.1 with cuda 12.6）
3. The compilation process is extremely long, and memory usage exceeds 90% (60GB).


I attempted to compile and deploy TEI on a Jetson AGX Orin 64G and found that it could not recognize SM87. Therefore, I modified the compute_cap_matching() function in backends/candle/src/compute_cap.rs to add support for the SM87 environment and architecture. Such modifications may not be effective in many cases, but fortunately, after making these changes, I was able to achieve support on the Jetson AGX Orin 64G. Not only did it not produce any errors, but it also showed excellent performance.

``` bash 
curl 127.0.0.1:8080/rerank \
    -X POST \
    -d '{"query": "What is Deep Learning?", "texts": ["Deep Learning is not...", "Deep learning is..."]}' \
    -H 'Content-Type: application/json'
```

logs

``` bash
2025-01-04T20:38:18.706787Z  INFO text_embeddings_backend_candle: backends/candle/src/lib.rs:292: Starting FlashBert model on Cuda(CudaDevice(DeviceId(1)))
2025-01-04T20:38:31.539445Z  INFO text_embeddings_router: router/src/lib.rs:248: Warming up model
2025-01-04T20:38:32.189069Z  INFO text_embeddings_router::http::server: router/src/http/server.rs:1812: Starting HTTP server: 0.0.0.0:8080
2025-01-04T20:38:32.189098Z  INFO text_embeddings_router::http::server: router/src/http/server.rs:1813: Ready
2025-01-04T20:44:11.047170Z  INFO rerank{total_time="177.15121ms" tokenization_time="727.783µs" queue_time="79.024583ms" inference_time="87.618256ms"}: text_embeddings_router::http::server: router/src/http/server.rs:459: Success
```

More screenshots are as follows:
![74f61043ee2b7a9da75e93547f397b7](https://github.com/user-attachments/assets/04a0ad86-6467-4cd5-a773-f0b8ba437e42)
![6519f72d3f88800fd8e89705b2d5fe1](https://github.com/user-attachments/assets/96be9873-67b9-4b86-96b5-5596d4a93055)
![fc2f46532bbe2545f4f7f2c55388190](https://github.com/user-attachments/assets/327e5227-7d98-426b-a45e-db37580a460e)


Therefore, I created a branch and added test code. After testing it in my application, I submitted a merge request.
Additionally, I also tried other embedding and rerank models, which ran well.

r0kk commented:
@HiddenPeak 
I can confirm that this is working on `Jetson AGX 64GB`. Thank you very much 🙏.

HiddenPeak commented:
> @HiddenPeak I can confirm that this is working on `Jetson AGX 64GB`. Thank you very much 🙏.

It's very cool~

taresh18-ag commented:
Hi, great work.

How did you get it running on jetson orin? When I try to compile it, it throws this error

![image](https://github.com/user-attachments/assets/d122422d-61a6-4415-a58e-276a595e84d5)

these are the steps I followed:

curl https://sh.rustup.rs/ -sSf | sh
sudo apt-get install libssl-dev gcc -y
git clone https://github.com/huggingface/text-embeddings-inference.git
cd text-embeddings-inference
cargo install --path router -F candle-cuda -F http --no-default-features # getting error here

thomas-hiddenpeak commented:
add -F dynamic-linking

r0kk commented:
Following process worked for me:

### Add `NVCC` to the path

NVIDIA's NVCC (NVIDIA CUDA Compiler) is a compiler driver used to compile CUDA (Compute Unified Device Architecture) code, which allows developers to write programs that run on NVIDIA GPUs. It translates CUDA code into executable binaries for GPU acceleration.

1. Check if `nvcc` exists
```bash
ls /usr/local/cuda/bin/nvcc
```

2. Update `.env` variables

    - Open **.bashrc**
        ```bash
        nano ~/.bashrc
        ```

    - add `nvcc` paths
        ```bash
        export PATH=/usr/local/cuda/bin:$PATH
        export LD_LIBRARY_PATH=/usr/local/cuda/lib64:$LD_LIBRARY_PATH
        ```

    - restart terminal
        ```bash
        source ~/.bashrc
        ```

    - check `nvcc` version
        ```bash
        nvcc --version
        ```

### Build Process (you can skip if build exists)
We prepare the build and can be found in current repository. If it doesn't exist, you can follow instructions below:

1. Install `Rust`
    ```bash
    curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh
    source $HOME/.cargo/env
    ```

2. Clone the Project
It is the branch of the original project, because at the time of the writing official release for `Jetson` family didn't exist.
    ```bash
    git clone https://github.com/HiddenPeak/text-embeddings-inference.git
    cd text-embeddings-inference
    ```

3. Install `ssl` (sometimes openssl problem might appear when building)
    ```bash
    sudo apt install libssl-dev
    ```

4. Build
 - move inside router dir inside the project
 - to use less space on `Jetson`, set `--target-dir` to external disc
    ```bash
    cd router/
    cargo build --release --features=candle-cuda --target-dir <target dir for generated artifact>
    ```

thomas-hiddenpeak committed:
feat: SM87/SM89 adaptation and community PRs integration

This commit adapts text-embeddings-inference for NVIDIA Jetson Orin (SM87)
and L4 GPU (SM89), and integrates valuable community PRs.

Changes:

1. SM87/SM89 CUDA Support
   - Added compute capability 8.7 and 8.9 support
   - Modified Dockerfile-cuda-all for multi-arch builds
   - Updated compute_cap.rs for SM87/89 detection
   Files: Dockerfile-cuda-all, cuda-all-entrypoint.sh, compute_cap.rs

2. PR #730: Qwen3 Reranker Support
   - Added classification head for Qwen3 reranking
   - Implemented template formatting system for chat-based reranking
   Files: models/qwen3.rs, core/templates.rs, core/lib.rs

3. PR #787: Batch Notification Performance Optimization
   - Implemented AtomicUsize counter for batch processing
   - Reduced unnecessary notify_one() calls
   - Only last request in batch triggers thread notification
   Files: core/infer.rs, router/http/server.rs, router/grpc/server.rs

4. PR #753: GeLU Activation Consistency Fix
   - Changed Gelu from approximate (gelu) to exact (gelu_erf)
   - Added NewGelu variant for backward compatibility
   Files: layers/linear.rs

5. PR #790: StaticEmbedding Model Support
   - Added support for 0_StaticEmbedding/ directory structure
   - Implemented fallback loading for model weights and tokenizer
   - Default to Mean pooling for StaticEmbedding models
   Files: models/static_embedding.rs (new), lib.rs, download.rs, router/lib.rs

6. PR #746: DebertaV2 Sequence Classification Support
   - Complete DebertaV2 model implementation
   - Support for sequence classification tasks (e.g., Llama Prompt Guard)
   - CPU and CUDA device support
   Files: models/debertav2.rs (new), lib.rs, models/mod.rs

All changes have been tested and compile successfully with:
  cargo check --all-targets

Compilation verified with CUDA support:
  cargo install --path router -F candle-cuda

Target Hardware: NVIDIA Jetson Orin AGX (SM87), L4 GPU (SM89)
Date: January 5, 2026

thomas-hiddenpeak committed:
Add SM 110 (Blackwell) compute capability support

thomas-hiddenpeak committed:
Merge branch 'main' into Jetson-support
